parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java

```diff
@@ -109,7 +110,12 @@ public BytesInput decompress(BytesInput bytes, int uncompressedSize) throws IOEx
           decompressor.reset();
         }
         InputStream is = codec.createInputStream(bytes.toInputStream(), decompressor);
-        decompressed = BytesInput.from(is, uncompressedSize);
+        if (codec instanceof ZstandardCodec) {
```

Comment:
This looks a little weird, but considering doing so will load the decompressor stream into heap in advance, and only zstd has this problem currently, so I made this modification only for zstd stream.

Reply:
Maybe we can consider closing the decompressed stream after it has been read:
https://github.com/apache/parquet-mr/blob/0819356a9dafd2ca07c5eab68e2bffeddc3bd3d9/parquet-common/src/main/java/org/apache/parquet/bytes/BytesInput.java#L283-L288
But I'm not sure if there is a situation where the decompressed stream is read more than once.

Reply:
The change looks OK to me, we probably should add some comments explaining why ZSTD deserves the special treatment here. 

The change on `BytesInput` looks more intrusive since it is used not only for decompression but other places like compression. For instance, `BytesInput.copy` calls `toByteArray` underneath, and after the call the original object should still be valid.

Reply:
Added comment.